### PR TITLE
Replace validator wait for activation stream with polling

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

deprecates the wait for activation event stream. This PR simplifies the checks for validator activation by polling against the status every epoch instead of receiving gRPC streams. By implementing this way we move one step closer to migrating to a full REST based validator client that can more easily connect with beacon nodes from other clients. 

tested with web UI in holesky

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](https://github.com/prysmaticlabs/prysm/blob/develop/CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: validator/client/validator_test.go [hunk 1]
```diff
@@ -341,46 +327,6 @@ func TestCanonicalHeadSlot_OK(t *testing.T) {
 	assert.Equal(t, primitives.Slot(0), headSlot, "Mismatch slots")
 }
 
-func TestWaitMultipleActivation_LogsActivationEpochOK(t *testing.T) {
```
Comment (james-prysm) [inline validator/client/validator_test.go:344]: duplicate test no longer needed, covered in activation_test

File: validator/client/wait_for_activation.go [hunk 2]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+	}
+	if len(valCounts) > 0 {
+		valCount = int64(valCounts[0].Count)
+	}
+	return valCount, nil
+}
+
+func (v *validator) handleReconnection(ctx context.Context, span octrace.Span, err error, message string, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	tracing.AnnotateError(span, err)
+	attempts := streamAttempts(ctx)
```
Comment (james-prysm) [inline validator/client/wait_for_activation.go:97]: should i rename this streamAttempts variable?

File: CHANGELOG.md [hunk 3]
```diff
@@ -53,6 +53,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 ### Deprecated
 - `--disable-grpc-gateway` flag is deprecated due to grpc gateway removal.
 - `--enable-experimental-state` flag is deprecated. This feature is now on by default. Opt-out with `--disable-experimental-state`.
+- `/eth/v1alpha1/validator/activation/stream` grpc wait for activation stream
```
Comment (saolyn) [inline CHANGELOG.md:56]: Nit
```suggestion
- `/eth/v1alpha1/validator/activation/stream` grpc wait for activation stream is depracated.
```

File: validator/client/wait_for_activation.go (hunk 2 above)
Comment (saolyn) [reply validator/client/wait_for_activation.go:97]: The code block is  small enough that I don't think it matters too much. It should be very clear what it is

File: validator/client/wait_for_activation.go [hunk 4]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
-			if err != nil && !errors.Is(err, iface.ErrNotSupported) {
-				return errors.Wrap(err, "could not get active validator count")
-			}
-
-			if len(valCounts) > 0 {
-				valCount = int64(valCounts[0].Count)
-			}
-
-			someAreActive = v.checkAndLogValidatorStatus(valCount)
+func (v *validator) waitForAccountsChange(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
```
Comment (saolyn) [inline validator/client/wait_for_activation.go:104]: We have some of these unexported functions with comments and others not. Should we just add comments to the rest of them?

File: validator/client/wait_for_activation.go [hunk 5]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
 		}
+		if err := v.waitForNextEpoch(ctx, headSlot, v.genesisTime, accountsChangedChan); err != nil {
+			return v.handleReconnection(ctx, span, err, "Failed to wait for next epoch. Reconnecting...", accountsChangedChan)
+		}
+		return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
 	}
+}
 
-	return nil
+// WaitForNextEpoch creates a blocking function to wait until the next epoch start given the current slot
```
Comment (saolyn) [inline validator/client/wait_for_activation.go:136]:
```suggestion
// waitForNextEpoch creates a blocking function to wait until the next epoch start given the current slot
```

File: validator/client/wait_for_activation.go (hunk 2 above)
Comment (james-prysm) [reply validator/client/wait_for_activation.go:97]: i guess it's not really a stream

File: validator/client/wait_for_activation.go (hunk 4 above)
Comment (james-prysm) [reply validator/client/wait_for_activation.go:104]: i don't think comments are required on unexported, it's probably fine to keep whatever we have.

File: validator/client/wait_for_activation.go [hunk 6]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+}
 
-	return nil
+// waitForNextEpoch creates a blocking function to wait until the next epoch start given the current slot
+func (v *validator) waitForNextEpoch(ctx context.Context, slot primitives.Slot, genesisTimeSec uint64, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	firstSlotOfNextEpoch, err := slots.EpochStart(slots.ToEpoch(slot) + 1)
+	if err != nil {
+		return err
+	}
+	nextEpochStartDuration, err := slots.ToTime(genesisTimeSec, firstSlotOfNextEpoch)
```
Comment (rkapka) [inline validator/client/wait_for_activation.go:142]: nitpick: the result of `ToTime` is not a duration, but a point in time

File: validator/client/wait_for_activation.go [hunk 7]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+	if err != nil {
+		return err
+	}
+	ss, err := slots.SecondsSinceSlotStart(slot, genesisTimeSec, uint64(time.Now().Unix()))
+	if err != nil {
+		return err
+	}
+	waitTime := uint64(nextEpochStartDuration.Unix()-time.Now().Unix()) + ss
```
Comment (rkapka) [inline validator/client/wait_for_activation.go:150]: I think this is incorrect, you should not be adding `ss`. You already know the exact point in time when the next epoch will start, so subtracting the current time from that time gives you the correct answer.

Comment (james-prysm) [reply validator/client/wait_for_activation.go:150]: I tested it, the time isn't always 100% exact, sometimes it's like 1-2 seconds off when the new epoch starts
the +ss is just from the current time in the slot so it guarantees it is within the bounds of the start of the first epoch. sometimes adding a 1 second buffer doesn't work

File: validator/client/wait_for_activation.go [hunk 8]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+	log.WithFields(logrus.Fields{
+		"slot":                   slot,
+		"seconds_sinceStart":     ss,
+		"next_epoch_start_slot":  firstSlotOfNextEpoch,
+		"slots_until_next_start": firstSlotOfNextEpoch - slot,
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/wait_for_activation.go:155]: it was useful for debugging, just having the slot wasn't always meaningful/exact

File: validator/client/wait_for_activation.go [hunk 9]
```diff
@@ -40,107 +39,122 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+	// Step 5: Check and log validator statuses.
+	someAreActive := v.checkAndLogValidatorStatus(valCount)
+	if !someAreActive {
+		// Step 6: If no active validators, wait for accounts change, context cancellation, or next epoch.
+		return v.waitForActivationRetry(ctx, span, accountsChangedChan)
+	}
+	return nil
+}
+
+// getValidatorCount is an api call to get the current validator count
```
Comment (rkapka) [inline validator/client/wait_for_activation.go:79]:
```suggestion
// getValidatorCount is an api call to get the current validator count.
```
Currently the comment is displayed as
```
getValidatorCount is an api call to get the current validator count "-1" indicates that validator count endpoint is not supported by the beacon node.
```

File: validator/client/wait_for_activation.go [hunk 10]
```diff
@@ -40,107 +39,122 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+	log.WithError(err).WithField("attempts", attempts).Error(message)
+	// Reconnection attempt backoff, up to 60s.
+	time.Sleep(time.Second * time.Duration(math.Min(uint64(attempts), 60)))
+	return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
+}
+
+func (v *validator) waitForAccountsChange(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	select {
+	case <-ctx.Done():
+		log.Debug("Context closed, exiting fetching validating keys")
```
Comment (rkapka) [inline validator/client/wait_for_activation.go:107]:
```suggestion
		log.Debug("Context closed, exiting waitForAccountsChange")
```
For uniformity with other logs

File: time/slots/slottime.go [hunk 11]
```diff
@@ -286,3 +287,27 @@ func WithinVotingWindow(genesisTime uint64, slot primitives.Slot) bool {
...
+	es := nextEpochStartTime.Unix()
+	n := time.Now().Unix()
+	waitTime := uint64(es - n)
+	log.WithFields(logrus.Fields{
+		"current_slot":           currentSlot,
+		"next_epoch_start_slot":  firstSlotOfNextEpoch,
+		"slots_until_next_start": firstSlotOfNextEpoch - currentSlot,
+		"total_wait_time":        waitTime,
+		"is_epoch_start":         IsEpochStart(currentSlot),
+	}).Warn("Waiting until next epoch before re-checking validator statuses...")
```
Comment (rkapka) [inline time/slots/slottime.go:311]: This log does not belong here

File: time/slots/slottime.go [hunk 12]
```diff
@@ -286,3 +287,27 @@ func WithinVotingWindow(genesisTime uint64, slot primitives.Slot) bool {
...
+	}
+	es := nextEpochStartTime.Unix()
+	n := time.Now().Unix()
+	waitTime := uint64(es - n)
+	log.WithFields(logrus.Fields{
+		"current_slot":           currentSlot,
+		"next_epoch_start_slot":  firstSlotOfNextEpoch,
+		"slots_until_next_start": firstSlotOfNextEpoch - currentSlot,
+		"total_wait_time":        waitTime,
+		"is_epoch_start":         IsEpochStart(currentSlot),
```
Comment (rkapka) [inline time/slots/slottime.go:310]: Our convention is to use `camelCase` for log fields

**Review by rkapka — CHANGES_REQUESTED** [review]

File: validator/client/wait_for_activation.go [hunk 13]
```diff
@@ -40,107 +39,122 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+}
 
-	return nil
+// waitForNextEpoch creates a blocking function to wait until the next epoch start given the current slot
+func (v *validator) waitForNextEpoch(ctx context.Context, genesisTimeSec uint64, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	waitTime, err := slots.SecondsUntilNextEpochStart(genesisTimeSec)
+	if err != nil {
+		return err
+	}
+	log.WithField("seconds_until_next_epoch", waitTime).Warn("No active validator keys provided. Waiting until next epoch to check again...")
```
Comment (rkapka) [inline validator/client/wait_for_activation.go:137]:
```suggestion
	log.WithField("secondsUntilNextEpoch", waitTime).Warn("No active validator keys provided. Waiting until next epoch to check again...")
```

File: validator/client/wait_for_activation.go [hunk 14]
```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
...
+		log.Debug("Context closed, exiting fetching validating keys")
+		return ctx.Err()
+	case <-accountsChangedChan:
+		// Accounts (keys) changed, restart the process.
+		return v.internalWaitForActivation(ctx, accountsChangedChan)
+	default:
+		log.Warn("No active validator keys provided. Waiting until next epoch to check again...")
+		headSlot, err := v.CanonicalHeadSlot(ctx)
+		if err != nil {
+			return v.handleReconnection(ctx, span, err, "Failed to get head slot. Reconnecting...", accountsChangedChan)
```
[the remark this answers is not in the document]
Comment (james-prysm) [reply validator/client/wait_for_activation.go]: yes this should essentially restart the process

**Review by rkapka — APPROVED** [review]

**Pull request merged**